### DISCUSSION TITLE
Allow customization of built-in IDOM client

### DISCUSSION THREAD

**Discussion by rmorshea** [idom-team/idom#485] [Ideas]
It would be great if it were possible to do some basic templating on the IDOM client web page. Without this, if you want to include custom styles or JS in the `<head />` of the HTML markup you have to completely re-create the client.

The alternative I suppose would be to make it easier to create custom clients. I'm not entirely sure what that would look like though. Maybe just a cookie cutter repo? The downside there of course is that you can't automatically update the repo if you update the template.

**Comment by rmorshea** [comment]
This came up again: https://github.com/idom-team/idom/discussions/400

**Comment by rmorshea** [comment]
@brentbaum The least brittle option I've been able to think of is to try and load a `/custom.js` script in `index.html`:

```html
<script>
  import("/custom.js")
    .then(() => console.log("Loaded custom.js script"))
    .catch(() => console.log("No custom.js script found"));
</script>
```

This would allow you to serve a `custom.js` file that could manipulate the page in any way you might want:

```python
from tempfile import NamedTemporaryFile

from sanic import Sanic

import idom
from idom.server.sanic import PerClientStateServer

temp = NamedTemporaryFile()
open(temp.name, "w+").write("document.title = 'Hello World';")  # set the page title on load

app = Sanic()
app.static("/custom.js", temp.name, name="custom_js", content_type="text/javascript")

HelloWorld = idom.component(lambda: idom.html.h1("Hello World"))
PerClientStateServer(HelloWorld, app=app).run("127.0.0.1", 8000)
```

Thoughts on this approach?

**Comment by rmorshea** [comment]
I suppose some of these details could be abstracted away by making this a `server_config` option:

```python
import idom

idom.run(MyComponent, server_config={"custom_js": "path/to/custom.js"})
```

**Comment by brentbaum** [comment]
I like the second approach quite a bit! The killer here is that React doesn't allow you to render and execute <script> tags by default - I tried for an hour yesterday to get a simple `<script>alert("hi!")</script>` example to run before remembering. If I could run js inline the same way I can define inline css this wouldn't be a problem.

**Comment by rmorshea** [comment]
@brentbaum what do you think about making something like this into a component that you'd just include in your layout?

```python
from idom import Javascript

my_script = Javascript("path/to/my/script.js")

@idom.component
def MyApp():
    return idom.html.div(my_script, ...)
```

**Comment by rmorshea** [comment]
IDOM now has a [`<script />`](https://github.com/idom-team/idom/blob/2d52bb63ed896a9f7862aa8ecff2232dd384f0c7/src/idom/html.py#L253) element that can be used for this purpose.

**Comment by acivitillo** [comment]
OK I have been thinking about this too, currently I load tailwind CSS like this:

```python
html.div(html.link({"href": "/static/tailwind.css", "rel": "stylesheet"})),
```

Not ideal. This should be in `<head>`. So if I understand your solution, I'd create a `<script>` component to load some js which would, in turn, modify `<head>` to add a custom CSS link tag?

```python
from idom import Javascript

js = """var head= document.getElementsByTagName('head')[0];
        var link = document.createElement('link');
        link.type = 'text/css';
        link.rel = 'stylesheet';
        link.href = '/static/tailwind.css';
        head.appendChild(link);
"""

@idom.component
def MyApp():
    return idom.html.script(content=js)
```

**Reply by rmorshea** [reply]
The main purpose of `<head/>` is for consumers of the webpage to be able to read static metadata about the page without having to execute any JS (e.g. [Social Cards](https://developers.facebook.com/docs/sharing/webmasters#markup)) so appending metadata to the doc dynamically once it's loaded with JS doesn't really serve that purpose.

For your particular use case, I'd stick with the `html.link(...)` solution since that doesn't strictly require that the CSS be statically declare. For other use cases, what's really needed here is a templating solution for the client's HTML. The main use case for the new `html.script` element with respect to client customization would be to do things like change the pages title by setting `document.title`.

A note on the script element: as per [the docs](https://idom-docs.herokuapp.com/docs/_auto/apis.html#idom.html.script), the script body is also slightly different from a normal JS script - it should evaluate to a function:

```javascript
// this is run on every render
() => {
  // this is run once the script is loaded and each time its content changes
  return () => {
    // this is run when the script is unloaded (i.e. it's removed from the tree) or just before its content changes
  }
}
```

This is pretty similar to being able to adding a `use_effect` hook except that it runs client-side.

**Reply by rmorshea** [reply]
I think I'll update the behavior of the script to allow for the way you expected it to work though (i.e. as a normal JS script).

**Reply by Archmonger** [reply]
To add on to what Ryan said, serving all of your CSS in head is bad for another reason. You generally want to defer CSS/JS load until they are needed for the sake of performance.

Also @rmorshea , I think it should behave as a normal script by default and maybe have a parameter for modifying the behavior?

**Reply by rmorshea** [reply]
Agreed. Instead of assuming the script is like an effect, we should be able to check if the script evals to a function, and if it does, we'll call it like an effect. Could you write up an issue for that?

**Reply by Archmonger** [reply]
Will do.

Also, does the current script accept the `src` attribute? If not we should probably support that as well.

**Reply by rmorshea** [reply]
Hmm, hadn't considered that. Yeah, I'll probably want to rework the script a bit to make it map into a normal `<script/>` tag better.

**Reply by acivitillo** [reply]
I agree with all your points, we could document that for CSS and other things that go into `<head>` users should create a `@component`. I would be happy to help document these practices.
